coordinator/src/main/scala/filodb.coordinator/ShardManager.scala

```diff
@@ -508,59 +518,66 @@ private[coordinator] final class ShardManager(settings: FilodbSettings,
-    /* If no shards are assigned to a coordinator, no commands are sent. */
-    logger.info(s"Sending start ingestion message for dataset=$dataset to coordinator $coord for shards $shards")
-    for {shard <- shards} coord ! StartShardIngestion(dataset, shard, None)
   }
 
-  /**
-    * This method has the shared logic for sending shard un-assignment messages
-    * to the coordinator, updating state for the event and broadcast of the state change to subscribers
-    */
-  private def sendUnassignmentMessagesAndEvents(dataset: DatasetRef,
-                                                coordinator: ActorRef,
-                                                mapper: ShardMapper,
-                                                shardsToDown: Seq[Int],
-                                                nodeUp: Boolean = true): Unit = {
-    logger.info(s"Sending stop ingestion message for dataset=$dataset to " +
+  private def doUnassignShards(dataset: DatasetRef,
+                               coordinator: ActorRef,
+                               shardsToDown: Seq[Int]): Unit = {
+    logger.info(s"Unassigning shards for dataset=$dataset to " +
       s"coordinator $coordinator for shards $shardsToDown")
     for { shard <- shardsToDown } {
       val event = ShardDown(dataset, shard, coordinator)
       updateFromShardEvent(event)
-      if (nodeUp) coordinator ! StopShardIngestion(dataset, shard)
     }
-    publishSnapshot(dataset)
-    logger.debug(s"Unassigned shards $shardsToDown of dataset=$dataset from $coordinator")
+  }
+
+  /**
+    * To be called after making a bunch of changes to the ShardMapper for the given dataset.
+    * Calling this method more often is permitted, but it generates more publish messages
+    * than is necessary.
+    */
+  private def publishChanges(ref: DatasetRef): Unit = {
+    publishSnapshot(ref)
+    updateShardMetrics()
   }
 
   /** Publishes a ShardMapper snapshot of given dataset to all subscribers of that dataset. */
-  def publishSnapshot(ref: DatasetRef): Unit =
+  def publishSnapshot(ref: DatasetRef): Unit = {
     mapperOpt(ref) match {
-      case Some(snapshot) =>
+      case Some(snapshot) => {
         for {
           subscription <- _subscriptions.subscription(ref)
         } subscription.subscribers foreach (_ ! snapshot)
+
+        // Also send a complete state command to all ingestion actors. Most are expected to
+        // actually be ingesting something, so also send the dataset info. This might be
+        // redundant, but sending it again is harmless.
+
+        val state = _datasetInfo(snapshot.ref)
+        val setupMsg = client.IngestionCommands.DatasetSetup(state.dataset.asCompactString,
+          state.storeConfig, state.source, state.downsample)
+
+        // TODO: Need to provide a globally consistent version, incremented when anything
+        //       changes, for any dataset.
+        val resync = ShardIngestionState(0, snapshot.ref, snapshot.map)
```

Comment:
Hmmm.   Seems like this will always send a 0 version....

Reply:
Yes. See TODO comment.